lib/exercism/user_finished_tracks.rb

```diff
@@ -0,0 +1,63 @@
+require_relative '../../x'
+class UserFinishedTracks
+  attr_reader :user
+  def initialize(user)
+    @user = user
+  end
+
+  # Return Hahs with, finished tracks
+  def list
+    @list ||= completed_tracks
+  end
+
+  def any?
+    !list.empty?
+  end
+
+  private
+
+  def completed_tracks
+    decorated_db_info.select do |hash|
+      hash['completed_count'].to_i == hash['total_in_track']
+    end
+  end
+
+  # Append more information, to the data extracted from the db.
+  # Example:
+  # [
+  #   {
+  #     'track_id' => 'go', 'completed_exercise' => 12,
+  #     'problems_count' => 30, 'language' => 'Go'
+  #   },
+  #   { 'track_id' => 'ruby', 'completed_exercise' => 60,
+  #     'problems_count' => 70, 'language' => 'Ruby'
+  #   }
+  # ]
+  def decorated_db_info
+    completed_exercises_by_language_and_user.each do |hash|
+      hash['total_in_track'] = track_problems_count[hash['track_id']]['problems_count']
+      hash['language'] = track_problems_count[hash['track_id']]['language']
+      hash
+    end
+  end
+
+  def completed_exercises_by_language_and_user
+    ActiveRecord::Base.connection.execute(completed_count_sql).to_a
+  end
+
+  def completed_count_sql
+    <<-SQL
+    SELECT language AS track_id, CAST (COUNT(id) AS INTEGER) AS completed_count FROM user_exercises
+    WHERE user_id=#{user.id}
+    AND (iteration_count > 0 OR skipped_at IS NOT NULL)
+    GROUP BY language
+    SQL
+  end
+
+  def track_problems_count
```

Comment:
We don't ever need to show the counts to the user in the view. Instead of constructing these hashes, how about using the database results to filter the `::X::Track.all` list?

```
X::Track.all.select { |track|
  track.problems.count <= started_tracks['solution_count']
}
```